File: src/domain/services/validator.py
import re

from mbs.domain.exceptions import ValidationError


class ValidatorService:

    def __init__(self):
        self.article_content_max_length = 32000
        self.article_title_max_length = 255
        self.article_title_min_length = 8
        self.article_content_min_length = 32

        self.comment_content_max_length = 1000
        self.comment_content_min_length = 8

        self.tag_title_max_length = 32
        self.tag_title_min_length = 3

        self.per_page_limit = 100

        self.filename_max_length = 255

        self.mime_type_max_length = 255  # RFC 4288 and RFC 6838
        self.mime_type_regex = re.compile(r"\w+/[-+.\w]+")

    def validate_article_title(self, title: str):
        if len(title) < self.article_title_min_length:
            raise ValidationError(f"Заголовок не может содержать меньше {self.article_title_min_length} символов")
        if len(title) > self.article_title_max_length:
            raise ValidationError(f"Заголовок не может содержать больше {self.article_title_max_length} символов")

    def validate_article_content(self, content: str):
        if len(content) < self.article_content_min_length:
            raise ValidationError(f"Публикация не может содержать меньше {self.article_content_min_length} символов")
        if len(content) > self.article_content_max_length:
            raise ValidationError(f"Публикация не может содержать больше {self.article_content_max_length} символов")

    def validate_comment_content(self, content: str):
        if len(content) < self.comment_content_min_length:
            raise ValidationError(f"Комментарий не может содержать меньше {self.comment_content_min_length} символов")
        if len(content) > self.comment_content_max_length:
            raise ValidationError(f"Комментарий не может содержать больше {self.comment_content_max_length} символов")

    def validate_tag_title(self, title: str):
        if len(title) < self.tag_title_min_length:
            raise ValidationError(f"Тег не может содержать меньше {self.tag_title_min_length} символов")
        if len(title) > self.tag_title_max_length:
            raise ValidationError(f"Тег не может содержать больше {self.tag_title_max_length} символов")

    def validate_pagination(self, page: int, per_page: int):
        if page < 1:
            raise ValidationError("Номер страницы не может быть меньше 1")
        if page > 2147483646:
            raise ValidationError("Номер страницы не может быть больше 2^32")

        if per_page < 1:
            raise ValidationError("Неверное количество элементов на странице")
        if per_page > self.per_page_limit:
            raise ValidationError(f"Количество элементов на странице не может быть больше {self.per_page_limit}")

    def validate_filename(self, filename: str):
        if len(filename) > self.filename_max_length:
            raise ValidationError(f"Имя файла не может содержать больше {self.filename_max_length} символов")

    def validate_mime_content_type(self, content_type: str):
        if len(content_type) > self.mime_type_max_length:
            raise ValidationError(f"Тип содержимого не может содержать больше {self.mime_type_max_length} символов")
        if not self.mime_type_regex.match(content_type):
            raise ValidationError("Неверный тип содержимого")

File: src/domain/services/validator.rs
use std::ops::Div;
use crate::domain::error::{DomainError, ValidationError};
use crate::domain::models::{
    article::{ARTICLE_CONTENT_MAX, ARTICLE_TITLE_MAX, ARTICLE_TITLE_MIN},
    file::{FILE_MIME_TYPE_MAX, FILE_MIME_TYPE_REGEX, FILE_NAME_MAX},
    tag::{TAG_TITLE_MAX, TAG_TITLE_MIN},
    comment::COMMENT_CONTENT_MAX
};

pub fn validate_article_title(title: &str) -> Result<(), DomainError> {
    if title.is_empty() {
        return Err(DomainError::Validation(
            ("title".into(), ValidationError::InvalidEmpty)
        ));
    }
    if !(ARTICLE_TITLE_MIN <= title.len() && title.len() <= ARTICLE_TITLE_MAX) {
        return Err(DomainError::Validation(
            ("title".into(), ValidationError::InvalidRange(ARTICLE_TITLE_MIN as i64, ARTICLE_TITLE_MAX as i64))
        ));
    }
    Ok(())
}

pub fn validate_article_content(content: &str) -> Result<(), DomainError> {
    if content.is_empty() {
        return Err(DomainError::Validation(
            ("content".into(), ValidationError::InvalidEmpty)
        ));
    }
    if !(ARTICLE_TITLE_MIN <= content.len() && content.len() <= ARTICLE_CONTENT_MAX) {
        return Err(DomainError::Validation(
            ("content".into(), ValidationError::InvalidRange(1, ARTICLE_CONTENT_MAX as i64))
        ));
    }
    Ok(())
}

pub fn validate_article_tags(tags: &[String]) -> Result<(), DomainError> {
    for (i, tag) in tags.iter().enumerate() {
        if tag.is_empty() {
            return Err(DomainError::Validation(
                (format!("tags_{i}"), ValidationError::InvalidEmpty)
            ));
        }
        if !(TAG_TITLE_MIN <= tag.len() && tag.len() <= TAG_TITLE_MAX) {
            return Err(DomainError::Validation(
                (format!("tags_{i}"), ValidationError::InvalidRange(TAG_TITLE_MIN as i64, TAG_TITLE_MAX as i64))
            ));
        }
    }
    Ok(())
}

pub fn validate_comment_content(content: &str) -> Result<(), DomainError> {
    if content.is_empty() {
        return Err(DomainError::Validation(
            ("content".into(), ValidationError::InvalidEmpty)
        ));
    }
    if !(1 <= content.len() && content.len() <= COMMENT_CONTENT_MAX) {
        return Err(DomainError::Validation(
            ("content".into(), ValidationError::InvalidRange(1, COMMENT_CONTENT_MAX as i64))
        ));
    }
    Ok(())
}

pub fn validate_filename(filename: &str) -> Result<(), DomainError> {
    if filename.is_empty() {
        return Err(DomainError::Validation(
            ("filename".into(), ValidationError::InvalidEmpty)
        ));
    }
    if !(1 <= filename.len() && filename.len() <= FILE_NAME_MAX) {
        return Err(DomainError::Validation(
            ("filename".into(), ValidationError::InvalidRange(1, FILE_NAME_MAX as i64))
        ));
    }
    Ok(())
}

pub fn validate_mime_content_type(content_type: &str) -> Result<(), DomainError> {
    if content_type.is_empty() {
        return Err(DomainError::Validation(
            ("mime_type".into(), ValidationError::InvalidEmpty)
        ));
    }
    if !(1 <= content_type.len() && content_type.len() <= FILE_MIME_TYPE_MAX) {
        return Err(DomainError::Validation(
            ("content_type".into(), ValidationError::InvalidRange(1, FILE_MIME_TYPE_MAX as i64))
        ));
    }
    if !FILE_MIME_TYPE_REGEX.is_match(content_type) {
        return Err(DomainError::Validation(
            ("content_type".into(), ValidationError::InvalidRegex(FILE_MIME_TYPE_REGEX.to_string()))
        ));
    }
    Ok(())
}


pub fn validate_page(page: u32) -> Result<(), DomainError> {
    if !(1 <= page && page <= 1_000_000) {
        return Err(DomainError::Validation(
            ("page".into(), ValidationError::InvalidRange(1, 1_000_000))
        ));
    }
    Ok(())
}

pub fn validate_per_page(per_page: u8) -> Result<(), DomainError> {
    if !(1 <= per_page && per_page <= 100) {
        return Err(DomainError::Validation(
            ("per_page".into(), ValidationError::InvalidRange(1, 100))
        ));
    }
    Ok(())
}